sycl/include/CL/sycl/ONEAPI/reduction.hpp

```diff
@@ -238,6 +259,41 @@ class reducer {
 
   T getIdentity() const { return MIdentity; }
 
+  template <typename _T = T>
+  enable_if_t<HasSameTypeArgPlusEq<_T>::value, reducer &>
```

Comment:
Am I right in thinking that this just checks if `+=` is defined for `T`?

If so, I don't think that's sufficient.  As written, a slightly modified version of the test from https://github.com/intel/llvm-test-suite/pull/140 would be allowed to compile without error:
```c++
Error += test<class C, float2>(float2{}, std::multiplies<>{});
```

I think we might need this to be something like:
```c++
enable_if_t<HasSameTypeArgPlusEq<_T>::value &&
            IsReduPlus<_T, BinaryOperation>::value,
            reducer &>
```
...with an equivalent change for the other operators.